database/optionally.go

```diff
@@ -0,0 +1,130 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icinga-go-library/com"
+	"github.com/pkg/errors"
+)
+
+// Upsert upserts entities to the database.
+type Upsert interface {
+	// Stream bulk upserts the specified entities via NamedBulkExec.
+	// The upsert statement is created using BuildUpsertStmt with the
```

Comment:
If not explicitly specified, the upsert statement is created using BuildUpsertStmt with the first entity from the entities stream.